content/en/docs/plugins/developers_guide.md

```diff
@@ -0,0 +1,1081 @@
+	}
+	if _, ok := obj["start"]; !ok {
+		return nil, fmt.Errorf("params %s did not contain start property", prms)
+	}
+
+	if _, ok := obj["maxEvents"]; !ok {
+		return nil, fmt.Errorf("params %s did not contain maxEvents property", prms)
+	}
+
+	return &MyInstance{
+		initParams: prms,
+		maxEvents:  obj["maxEvents"],
+		counter:    0,
+		sample:     obj["start"],
+	}, nil
+}
+
+func (m *MyInstance) Close() {
+	log.Printf("[%s] Close\n", PluginName)
+}
+```
+
+#### Returning Events
+
+New events are generated in batch by the `NextBatch` function. The function is mandatory for source plugins and must be defined as a method of the plugin instance struct type. The `pState` argument is the plugin struct type initialized in `Init()`, passed in by the framework for ease of access. The plugin state is passed as an instance of the `sdk.PluginState` interface, so a manual cast is required to access the internal state variables defined in the struct type.
+
+The `evts` parameter is a sdk-managed batch of events to be used for creating new events. For that, the SDK uses a slab allocator and reuses the same event batch at every iteration to improve performance. The lenght of the `evts` list represents the maximum size of each event batch.
+Each element of the batch is an instance of `sdk.EventWriter` that provides handy methods to write the event info and data. Event data can be written with the Go `io.Writer` interface.
+
+If an error is returned, the SDK returns a failure to the framework and invalidates the current batch. The special errors `sdk.ErrTimeout` and `sdk.ErrEOF` have a special meaning, and are used to either advise the framework that no new events are currently available, or that the event stream is terminated.
+
+```go
+func (m *MyInstance) NextBatch(pState sdk.PluginState, evts sdk.EventWriters) (int, error) {
+	log.Printf("[%s] NextBatch\n", PluginName)
+
+	// Return EOF if reached maxEvents
+	if m.counter >= m.maxEvents {
+		return 0, sdk.ErrEOF
+	}
+
+	var n int
+	var evt sdk.EventWriter
+	myPlugin := pState.(*MyPlugin)
+	for n = 0; m.counter < m.maxEvents && n < evts.Len(); n++ {
+		evt = evts.Get(n)
+		m.counter++
+
+		// Increment sample by 1, also add a jitter of [0:jitter]
+		m.sample += 1 + uint64(myPlugin.rand.Int63n(int64(myPlugin.jitter+1)))
+
+		// The representation of a dummy event is the sample as a string.
+		str := strconv.Itoa(int(m.sample))
+
+		// It is not mandatory to set the Timestamp of the event (it
+		// would be filled in by the framework if set to uint_max),
+		// but it's a good practice.
+		evt.SetTimestamp(uint64(time.Now().UnixNano()))
+
+		_, err := evt.Writer().Write([]byte(str))
+		if err != nil {
+			return 0, err
+		}
+	}
+	return n, nil
+}
+```
+
+#### Printing Events As Strings
+
+Source plugins must define a `String()` method to format the contents of events created with a previous call to `NextBatch()`. The event data is available through an `io.ReadSeeker` interface. 
```

Comment:
I'd reiterate that we use the io.ReadSeeker interface to avoid data copies and keep the common paths fast.

Reply:
Looks great thx!